fdbserver/storageserver.actor.cpp

```diff
@@ -9109,408 +9176,426 @@ ACTOR Future<Void> fetchKeys(StorageServer* data, AddingShard* shard) {
 					rangeEnd = keys.end;
 				}
 			} else if (conductBulkLoad) {
-				ASSERT(dataMoveIdIsValidForBulkLoad(dataMoveId)); // TODO(BulkLoad): remove dangerous assert
-				// Get the bulkload task metadata from the data move metadata. Note that a SS can receive a data move
-				// mutation before the bulkload task metadata is persisted. In this case, the SS will not be able to
-				// read the bulkload task. SS will wait at this point until the bulkload task metadata is persisted.
-				// Moreover, the bulkload task metadata is persist at a verison at least the version when this SS
-				// receives the datamove mutation. Therefore, the SS should read the bulkload task metadata at a version
-				// at least this SS version.
-				// Note that it is possible that this SS can never get the bulkload metadata because the bulkload data
-				// move is cancelled or replaced by another data move. In this case, while
-				// getBulkLoadTaskStateFromDataMove get stuck, this fetchKeys is guaranteed to be cancelled.
-				BulkLoadTaskState bulkLoadTaskState = wait(getBulkLoadTaskStateFromDataMove(
-				    data->cx, dataMoveId, /*atLeastVersion=*/data->version.get(), data->thisServerID));
-				TraceEvent(bulkLoadVerboseEventSev(), "SSBulkLoadTaskFetchKey", data->thisServerID)
-				    .detail("DataMoveId", dataMoveId.toString())
-				    .detail("Range", keys)
-				    .detail("Phase", "Got task metadata");
-				// Check the correctness: bulkLoadTaskMetadata stored in dataMoveMetadata must have the same
-				// dataMoveId.
-				ASSERT(bulkLoadTaskState.getDataMoveId() == dataMoveId);
-				// We download the data file to local disk and pass the data file path to read in the next step.
-				localBulkLoadFileSets = std::make_shared<BulkLoadFileSetKeyMap>();
-				wait(bulkLoadFetchKeyValueFileToLoad(
-				    data, bulkLoadLocalDir, bulkLoadTaskState, /*output=*/localBulkLoadFileSets));
-				TraceEvent(bulkLoadVerboseEventSev(), "SSBulkLoadTaskFetchKey", data->thisServerID)
-				    .detail("DataMoveId", dataMoveId.toString())
-				    .detail("Range", keys)
-				    .detail("Phase", "File download");
-				hold = tryGetRangeForBulkLoad(results, keys, localBulkLoadFileSets);
+				// Call the helper actor for bulk load logic
+				state std::pair<bool, std::shared_ptr<BulkLoadFileSetKeyMap>> bulkLoadResult =
+				    wait(fetchKeysBulkLoad(data, fetchKeysID, bulkLoadLocalDir, dataMoveId, keys, results));
```

Comment:
This code refactoring has some issue.

The functionality of "hold" here is to keep `bulkLoadFetchKeyValueFileToLoad` running until it completes or the fetchKey completes. If we do with the new code, the  `bulkLoadFetchKeyValueFileToLoad` can stop without producing all key-value pairs because no memory holds the actor. 

With the new code, the actor `bulkLoadFetchKeyValueFileToLoad ` memory is hold by `fetchKeysBulkLoad` internally, which means if the ` wait(fetchKeysBulkLoad()` completes, no memory is holding `bulkLoadFetchKeyValueFileToLoad`.